### PR TITLE
[Intel MKL] Refactor MKL Eager from std::vector to std::hashmap for cleaner design

### PR DESCRIPTION
**Pull request by nammbash**
Refactoring the MKL Eager from std::vector to std::hash_map and improve performance along with a cleaner design.

Related to the PR: tensorflow/tensorflow#33540 (See comment section from @penpornk )

### CODE REVIEW

**Comment by nammbash** [comment]
@penpornk: Related Refactor PR to #33540

File: tensorflow/core/common_runtime/eager/mkl_eager_op_rewrite.cc [hunk 1]
```diff
@@ -33,18 +32,11 @@ class MklEagerOpRewrite : public EagerOpRewrite {
...
-    // Overload Operator== for std::find comparison
-    // used by SlowCheckIfKernelRegistered.
-    bool operator==(const MklEagerOp& rhs) const {
-      return (op_name.compare(rhs.op_name) == 0);
-    }
   };
 
  private:
-  // TODO(intel-tf): refactor with unordered_map;
   // especially when adding more ops/rewrite rules in future.
```
Comment (penpornk) [inline tensorflow/core/common_runtime/eager/mkl_eager_op_rewrite.cc]: Please delete this orphaned comment.

File: tensorflow/core/common_runtime/eager/mkl_eager_op_rewrite.cc [hunk 2]
```diff
@@ -69,49 +61,51 @@ class MklEagerOpRewrite : public EagerOpRewrite {
...
-  // This is called by FastCheckIfKernelRegistered once per unique op name and
-  // data type.
-  bool SlowCheckIfKernelRegistered(string op_name, DataType dt);
+  // Helper function to insert mkl_eager_ops to Map
+  void InsertMKLEagerOps(MklEagerOp op);
 
-  // map used by FastCheckIfKernelRegistered.
-  std::unordered_map<string, bool> registered_kernels_map;
+  // Map used by FastCheckIfKernelRegistered.
+  std::unordered_map<std::string, bool> registered_kernels_map;
```
Comment (penpornk) [inline tensorflow/core/common_runtime/eager/mkl_eager_op_rewrite.cc]: Nit: Class member naming.
```suggestion
  std::unordered_map<std::string, bool> registered_kernels_map_;
```
Sorry I didn't notice this in the last PR.

File: tensorflow/core/common_runtime/eager/mkl_eager_op_rewrite.cc [hunk 3]
```diff
@@ -237,9 +229,10 @@ bool MklEagerOpRewrite::SlowCheckIfKernelRegistered(string op_name,
 }
 
 Status MklEagerOpRewrite::RewriteToMklOp(
-    EagerOperation* orig_op, std::unique_ptr<EagerOperation>* mkl_op,
-    const int op_idx) {
-  mkl_eager_ops_[op_idx].CreateMklOp(orig_op, mkl_op);
+    EagerOperation* orig_op, std::unique_ptr<EagerOperation>* mkl_op) {
+  // TODO(intel-tf): mkl_eager_ops_ lookup can be reduced from twice
+  // (once each in ShouldRewriteOp & RewriteToMklOp) to just once.
```
Comment (penpornk) [inline tensorflow/core/common_runtime/eager/mkl_eager_op_rewrite.cc]: I was thinking about this too. Thank you for adding the TODO!

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you very much for the PR!

File: tensorflow/core/common_runtime/eager/mkl_eager_op_rewrite.cc (hunk 1 above)
Comment (nammbash) [reply tensorflow/core/common_runtime/eager/mkl_eager_op_rewrite.cc]: yes done!

File: tensorflow/core/common_runtime/eager/mkl_eager_op_rewrite.cc (hunk 2 above)
Comment (nammbash) [reply tensorflow/core/common_runtime/eager/mkl_eager_op_rewrite.cc]: no worries. My bad!

File: tensorflow/core/common_runtime/eager/mkl_eager_op_rewrite.cc (hunk 3 above)
Comment (nammbash) [reply tensorflow/core/common_runtime/eager/mkl_eager_op_rewrite.cc]: my pleasure!

**Comment by nammbash** [comment]
@penpornk All comments are addressed!

**Review by penpornk — APPROVED** [review]
Thank you again!

**Pull request merged**